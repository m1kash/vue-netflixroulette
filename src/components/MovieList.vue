<template>
  <div v-if="items.length" class="aqa-movie-list grid gap-12 gap-y-8 grid-cols-3 py-8">
  <movie-item v-for="({title, poster_path, id, release_date, genres}) in items"
              :title="title" :image="poster_path" :release_date="release_date"
              :genres="genres" :key="id" :id="id"  ></movie-item>
  </div>
  <div class="aqa-404 text-4xl text-white text-center mt-4 mb-16" v-else>
    Oh no ðŸ˜¢ 404
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from 'vue';
import IMovie from '@/types/IMovie';
import MovieItem from '@/components/MovieItem.vue';

export default defineComponent({
  name: 'movie-list',
  components: {
    MovieItem,
  },
  props: {
    items: {
      type: Array as PropType<IMovie[]>,
    },
  },
});
</script>
