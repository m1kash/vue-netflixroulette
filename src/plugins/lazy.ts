import { Directive, Plugin } from 'vue';

export interface IOptionsLazy {
  width: number,
  height: number,
  placeText: string,
  customImage?: string,
}

const lazyPlugin: Plugin = {
  install: (app, options: IOptionsLazy) => {
    const img = `https://via.placeholder.com/${options.width}x${options.height}.png?text=${options.placeText}`;
    const placeholder = options.customImage
      || img;

    app.directive('lazy', {
      mounted(el: HTMLElement, bindings) {
        const callback: IntersectionObserverCallback = function callbackLazy(
          entries,
        ) {
          el.setAttribute('src', placeholder);
          entries.forEach((entry) => {
            if (entry.isIntersecting) {
              el.setAttribute('src', bindings.value);
              el.addEventListener('error', () => {
                el.setAttribute('src', placeholder);
              });
              // eslint-disable-next-line
              lazyImageObserver.unobserve(el);
            }
          });
        };
        const lazyImageObserver = new IntersectionObserver(callback);

        lazyImageObserver.observe(el);
      },
    } as Directive);
  },
};

export default lazyPlugin;
