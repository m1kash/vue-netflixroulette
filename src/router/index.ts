import Home from '@/pages/Home.vue';
import Movie from '@/pages/Movie.vue';
import {createRouter, createWebHistory} from "vue-router";

const routes = [
  {
    path: '/movies/:id',
    component: Movie,
    name: 'Movies',
  },
  {
    path: '/',
    component: Home,
    name: 'Home'
  }
]

const router = createRouter({
  routes,
  linkActiveClass: 'active',
  history: createWebHistory(''),
  scrollBehavior () {
    return { top: 0 }
  }
});

export default router;
