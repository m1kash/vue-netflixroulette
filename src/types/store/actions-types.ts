import { ActionContext, ActionTree } from 'vuex';
import IState from '@/types/IState';
import { searchBy } from '@/types/searchTag';
import { sortBy } from '@/types/sortBy';

export enum ActionsTypes {
  FILTER_SEARCH_TEXT = 'FILTER_SEARCH_TEXT',
  FILTER_SEARCH_BY = 'FILTER_SEARCH_BY',
  CHANGE_SORT_BY = 'CHANGE_SORT_BY',
  GET_MOVIE_BY_ID = 'GET_MOVIE_BY_ID',
}

export interface IActions extends ActionTree<IState, IState> {
  [ActionsTypes.FILTER_SEARCH_TEXT](
    context: ActionContext<IState, IState>,
    payload: string):void
  [ActionsTypes.FILTER_SEARCH_BY](
    context: ActionContext<IState, IState>,
    payload: searchBy):void
  [ActionsTypes.CHANGE_SORT_BY](
    context: ActionContext<IState, IState>,
    payload: sortBy):void
  [ActionsTypes.GET_MOVIE_BY_ID](
    context: ActionContext<IState, IState>,
    payload: string):void
}
