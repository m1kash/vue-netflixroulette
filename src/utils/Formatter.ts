import IFormatter from "@/types/IFormatter";

const Formatter: IFormatter = {
  year(date) {
    const DateInstance = new Date(date) as Date;

    return DateInstance.getFullYear();
  },
  separator(items, separator) {
    return items.join(separator);
  },
  duration(min) {

    return `${min || 0} min`;
  }
};

export default Formatter;
