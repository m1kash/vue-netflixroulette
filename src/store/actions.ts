import {TActions, ActionsTypes} from "@/types/store/actions-types";
import {MutationTypes} from "@/types/store/mutation-types";

export const actions: TActions = {
  [ActionsTypes.FILTER_SEARCH_TEXT]({ commit }, payload) {
    commit(MutationTypes.SET_SEARCH_TEXT, payload);
    commit(MutationTypes.FILTER_MOVIES);
  },
  [ActionsTypes.FILTER_SEARCH_BY]({ commit }, payload) {
    commit(MutationTypes.SET_SEARCH_BY, payload);
    commit(MutationTypes.FILTER_MOVIES);
  },
  [ActionsTypes.CHANGE_SORT_BY]({ commit }, payload) {
    commit(MutationTypes.SET_SORT_BY, payload);
    commit(MutationTypes.FILTER_MOVIES);
  },
  [ActionsTypes.GET_MOVIE_BY_ID]({ commit }, payload) {
    commit(MutationTypes.SET_MOVIE, payload);
    commit(MutationTypes.FILTER_RELATED_MOVIES);
  },
}
