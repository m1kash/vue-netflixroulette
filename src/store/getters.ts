import { GetterTree } from 'vuex';
import IState from '@/types/IState';
import { Getters } from '@/types/store/getters-types';

const getters: GetterTree<IState, IState> & Getters = {
  getMovies(state) {
    return state.items;
  },
  getMovie(state) {
    return state.movie;
  },
  getRelatedMovies(state) {
    return state.relatedMovies;
  },
  getCountMovies(state) {
    return state.items.length;
  },
  getFirstGenreMovie(state) {
    return state.movie.genres ? state.movie.genres[0] : '';
  },
};

export default getters;
