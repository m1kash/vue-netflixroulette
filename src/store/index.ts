import IState from '@/types/IState';
import {createStore, Store, useStore as baseUseStore} from 'vuex'
import movies from '../../data/movies.json';
import { getters } from "@/store/getters";
import {mutations} from "@/store/mutations";
import {actions} from "@/store/actions";
import {InjectionKey} from "vue";

export const key: InjectionKey<Store<IState>> = Symbol();
const store = createStore<IState>({
  state () {
    return {
      items: movies,
      filteredMovies: [],
      relatedMovies: [],
      movie: {},
      sortBy: 'release_date',
      searchBy: 'title',
      searchText: ''
    }
  },
  getters: getters,
  mutations: mutations,
  actions: actions
});

export function useStore() {
  return baseUseStore(key);
}

export default store;
