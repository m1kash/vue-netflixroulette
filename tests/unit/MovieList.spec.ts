import 'jest';
import { shallowMount, mount, VueWrapper } from '@vue/test-utils';
import MovieList from "@/components/MovieList.vue";
import data from "../mock-data/movies.json";
import {createStore} from "vuex";
import {ActionsTypes} from "@/types/store/actions-types";
import {key} from "@/store";
import Formatter from "@/utils/Formatter";
import router from "@/router";
import lazy, {IOptionsLazy} from "@/plugins/lazy";
import setupIntersectionObserverMock from '../../config/test-utils/intersectionObserverMock';


describe('MovieList.vue', () => {
  let wrapper: VueWrapper;
  let store;
  let getMovieMock: jest.Mock;

  beforeEach(() => {
    setupIntersectionObserverMock();
    getMovieMock = jest.fn();
    store = createStore({
      actions: {
        [ActionsTypes.GET_MOVIE_BY_ID]: getMovieMock,
      }
    });
    wrapper = mount(MovieList, {
      props: {
        items: data
      },
      global: {
        provide: {
          [key]: store,
          'formatter': Formatter
        },
        plugins: [
          store,
          router,
          [lazy, {
            height: 592,
            width: 395,
            placeText: 'Loading...',
          } as IOptionsLazy]
        ],
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('render correct', () => {
    expect(wrapper.vm.$el).toMatchSnapshot();
  });

  test('should show need count movies equal count movies', () => {
    const movieList = wrapper.find('.aqa-movie-list');

    expect(movieList.element.children.length).toEqual(data.length);
  });

  test('movie clicked on button and keys', async () => {
    const ID_MOVIE = 1;
    const movieItem = wrapper.find('.aqa-movie');

    movieItem.trigger('click');
    await wrapper.vm.$nextTick();

    expect(getMovieMock).toBeCalledWith(expect.anything(), ID_MOVIE);
  });

  test('should show 404 when 0 movies showed', () => {
    const movieComponent = shallowMount(MovieList, {
      props: {
        items: []
      }
    });
    const movie404 = movieComponent.find('.aqa-404');
    expect(movie404.isVisible()).toBeTruthy();
  });
});
