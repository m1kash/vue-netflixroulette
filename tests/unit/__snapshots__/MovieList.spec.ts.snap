// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList.vue render correct 1`] = `
<div
  class="aqa-movie-list grid gap-12 gap-y-8 grid-cols-3 py-8"
>
  
  <div
    class="aqa-movie flex-col flex"
  >
    <img
      alt="image"
      height="370"
    />
    <div
      class="flex flex-wrap mt-2"
    >
      <div
        class="flex flex-col w-1/2"
      >
        <div
          class="text-zinc-500 text-lg"
        >
          Test 1
        </div>
      </div>
      <div
        class="w-1/2 flex justify-end items-start"
      >
        <div
          class="text-zinc-400 border border-solid border-zinc-500 rounded p-0.5 px-4"
        >
          2020
        </div>
      </div>
      <div
        class="text-zinc-600 w-full"
      >
        Action&Adventure&Science Fiction
      </div>
    </div>
  </div>
  <div
    class="aqa-movie flex-col flex"
  >
    <img
      alt="image"
      height="370"
    />
    <div
      class="flex flex-wrap mt-2"
    >
      <div
        class="flex flex-col w-1/2"
      >
        <div
          class="text-zinc-500 text-lg"
        >
          Test 2
        </div>
      </div>
      <div
        class="w-1/2 flex justify-end items-start"
      >
        <div
          class="text-zinc-400 border border-solid border-zinc-500 rounded p-0.5 px-4"
        >
          2019
        </div>
      </div>
      <div
        class="text-zinc-600 w-full"
      >
        Science Fiction&Action&Adventure
      </div>
    </div>
  </div>
  <div
    class="aqa-movie flex-col flex"
  >
    <img
      alt="image"
      height="370"
    />
    <div
      class="flex flex-wrap mt-2"
    >
      <div
        class="flex flex-col w-1/2"
      >
        <div
          class="text-zinc-500 text-lg"
        >
          Test 3
        </div>
      </div>
      <div
        class="w-1/2 flex justify-end items-start"
      >
        <div
          class="text-zinc-400 border border-solid border-zinc-500 rounded p-0.5 px-4"
        >
          2019
        </div>
      </div>
      <div
        class="text-zinc-600 w-full"
      >
        Action&Adventure&Science Fiction
      </div>
    </div>
  </div>
  
</div>
`;
